### PR TITLE
dotnet kernel for C# and F# 

### PR DESCRIPTION
**Pull request by anpin**
This PR adds dotnet kernel to C# and F# languages in jupyter notebooks. 

This PR depends on refactorings made in https://github.com/tweag/jupyenv/pull/524

Closes https://github.com/tweag/jupyenv/issues/393

### CODE REVIEW

**Comment by GTrunSec** [comment]
@anpin Thanks for your contributing. once you solve the CI to be green, I will merge your PR.

**Comment by GTrunSec** [comment]
After merging this PR, we should have a plan to migrate the poetry to the dream2nix. like https://github.com/hardenedlinux/AISecurity-Research-Template/blob/main/flake.nix

**Comment by anpin** [comment]
> @anpin Thanks for your contributing. once you solve the CI to be green, I will merge your PR.

removed dotnet-sdk from `requiredRuntimePackages` which fixed `nix flake check` locally

**Comment by anpin** [comment]
that's weird. can't reproduce these failures locally.

**Comment by GTrunSec** [comment]
@anpin  Did you run this command`nix build -L .#jupyterlab-kernel-example-dotnet-csharp` locally? since I can test them tomorrow night,  please check your test.py, and make sure everything goes well.

**Comment by anpin** [comment]
yes I do. There is no output in the case of a successful test, so it works fine. I don't have dotnet-sdk in my $PATH

image

**Comment by anpin** [comment]
added openssl and zlib to runtimeDeps, maybe that will fix CI

File: modules/kernels/dotnet/package.nix
Comment (anpin) [inline modules/kernels/dotnet/package.nix:14]: Eventually this should be upstreamed to nixpkgs and built from source

**Comment by anpin** [comment]
I just realized I have nix-ld on my machine. Maybe that’s why it doesn’t fail. Will try to run without it later.

**Comment by anpin** [comment]
I disabled nix-ld and reverted the last commit with runtimeDeps, but that did not make it fail. The kernel test executed successfully. 

I managed to use `lib.mapAttrs'` on the `example` set to run test.py in nixos flake check VM for every kernel. It seems to work for per-existing kernels, but those added in this PR fail with a different error than CI build above.

**Comment by anpin** [comment]
If running test.py in nix falke check is desired by maintainers I can submit separate a new PR or include here. Here is the gist of it

```diff
diff --git a/flake.nix b/flake.nix
index c5d35a5..3a4f4ed 100644
--- a/flake.nix
+++ b/flake.nix
@@ -133,6 +133,40 @@
 
         exampleJupyterlabAllKernelsNew =
           jupyterLib.mkJupyterlabNew (builtins.attrValues examples);
+
+        checkKernel = name: example:
+          with import (nixpkgs + "/nixos/lib/testing-python.nix") {inherit system;}; let
+            kernel = jupyterLib.mkJupyterlabNew example;
+            test_py = (builtins.dirOf example) + "/test.py";
+          in
+            makeTest {
+              inherit name;
+              nodes.machine = {config, ...}: {
+                config = {
+                  environment.systemPackages = [kernel];
+                };
+              };
+              testScript = ''
+                machine.execute("${kernel}/bin/python ${test_py}")
+              '';
+            };
+        exampleCheck = (
+          lib.mapAttrs'
+          (
+            name: value:
+              lib.nameValuePair
+              ("check-" + name)
+              (checkKernel name value)
+          )
+          examples
+        );
       in {
         lib =
           jupyterLib
@@ -163,9 +197,11 @@
             ${pre-commit.shellHook}
           '';
         };
-        checks = {
-          inherit pre-commit;
-        };
+        checks =
+          {
+            inherit pre-commit;
+          }
+          // exampleCheck;
         apps = {
           update-poetry-lock =
             flake-utils.lib.mkApp


```

**Review by GTrunSec — APPROVED** [review]
LGTM

**Comment by anpin** [comment]
thanks for merging it @GTrunSec! 

I've added another commit with above patch which allows to test any kernel via  `nix build .\#checks.x86_64-linux.check-example-dotnet-fsharp -L --show-trace`. Would you consider such change in a new PR?

**Pull request merged**
